experiment_prototype/scan_classes/sequences.py

```diff
@@ -337,107 +260,97 @@ def __init__(self, seqn_keys, sequence_slice_dict, sequence_interface, transmit_
+                phase_encoding = phase_encoding.reshape((phase_encoding.shape[0],-1))
 
-            repeat = one_pulse_list[0]['isarepeat']
-            timing = one_pulse_list[0]['pulse_timing_us']
-            pulse_samples = []
+                self.output_encodings[slice_id].append(phase_encoding)
 
+                # we have [p,e] and [a,s], but we want [p,a,(e*s)]. Adding null axis to encoding
+                # will produce this result.
+                phase_encoding = np.exp(1j * phase_encoding[:,np.newaxis,:])
+                samples = phase_encoding * basic_samples
 
-            if repeat:
-                pulse_antennas = []
             else:
-                # Initialize a list of lists for samples on all channels.
-                # TODO: modify this function if we put a weighting on powers instead of just a
-                # simple power_divider integer
-                pulse_samples, pulse_antennas = (
-                    make_pulse_samples(one_pulse_list, self.power_divider, self.slice_dict,
-                                       slice_to_beamdir_dict, txrate, txctrfreq,
-                                       main_antenna_count, main_antenna_spacing, pulse_ramp_time,
-                                       max_usrp_dac_amplitude, tr_window_time))
-                if pulse_index == 0:
-                    # calculate the first rx sample and set the value.
-                    self.first_rx_sample_time = calculate_first_rx_sample_time(
-                        pulse_samples[0].shape[0], txrate)
-                # Can plot for testing here
-                # plot_samples('channel0.png', pulse_samples[0])
-                # plot_fft('fftplot.png', pulse_samples[0], prog.txrate)
-
-            # This is all the data required for a pulse.
-            pulse_transmit_data['startofburst'] = startofburst
-            pulse_transmit_data['endofburst'] = endofburst
-            pulse_transmit_data['pulse_antennas'] = pulse_antennas
-            pulse_transmit_data['samples_array'] = pulse_samples
-            pulse_transmit_data['timing'] = timing
-            pulse_transmit_data['isarepeat'] = repeat
-
-            # Add pulse dictionary pulse_transmit_data at last place in sequence list
-            sequence_list.append(pulse_transmit_data)
-
-        self.find_blanks()
-
-        return sequence_list
+                samples = np.repeat(basic_samples[np.newaxis,:,:], num_pulses, axis=0)
+
+            # sum the samples into their position in the sequence buffer.
+            tr_window_num_samps = round(tr_window_time * txrate)
+            for i, pulse in enumerate(self.combined_pulses_metadata):
+                for component_info in pulse['component_info']:
+                    if component_info['slice_id'] == slice_id:
+                        pulse_timing_us = component_info['start_time_us']
+                        pulse_sample_start = round(txrate * (pulse_timing_us * 1e-6))
```

Comment:
so I'm thinking about this a bit more and all of these rounds are concerning me. At 5 MHz the difference between one sample to the next is 0.2 us. If we sample slower for any reason it will increase. Our timing accuracy will only be as good as the precision of indexing into the buffers. So we might want to put pulse_sample_number into the pulse dictionary instead of just start_time_us (do this in the init). Also, we could calculate the tr_window_num_samps in the init similar to how we are calculating number of receive samples there. Then we can check for the offsets due to digitizing there and perhaps log things or give warnings if it is over a certain amount. Might have to do this for both rx rate and tx rate since the rates can differ... or we can consider enforcing them to be the same...

Reply:
this will lead into the improvement of being able to specify the start sequence time in the init. If everything is based off of sample number in the init, the only time necessary will be the transmit and receive start times.

Reply:
also, notice you are using this pulse_sample_start value multiple times so if it's in the init it will reduce the number of calcs here.

Reply:
What do you mean by timing accuracy? And i can't see a way to avoid rounding at some point even if you go based off sample number, because at some point of converting we need to account for floating point precision issues. Also, I just want to remind that this part is only used for a place to combine samples. The driver/dsp performs a calculation to align the pulse transmit time to the start of a sample in the buffer.